### DISCUSSION TITLE
How to build a subproject which depends on user provided header files?

### DISCUSSION THREAD

**Discussion by crazybolillo** [mesonbuild/meson#11809] [Q&A] [answered]
I am trying to build the STM32CubeF4 libraries to program ST Micro-controllers. They provide everything as a github repository so I have [forked it and added a `meson.build` file to use it as a subproject](https://github.com/crazybolillo/STM32CubeF4).

Some of the files/libraries, specifically the HAL component, require the user to copy a [template header](https://github.com/crazybolillo/STM32CubeF4/blob/master/Drivers/STM32F4xx_HAL_Driver/Inc/stm32f4xx_hal_conf_template.h), rename it and use it to build the library.

This is making it very hard to actually use this as a subproject, since the library files depend on a file outside the subproject (a user provided file) and I can't use said file because it violates the sandbox.

Any ideas how to overcome this?

**Comment by eli-schwartz** [comment, marked as answer]
Does it have to be physically present in the subproject or is adding the header as an include_directories to the same target that also tries to depend on the subproject, work?

In the former case you could try adding the header file path as a subproject option, and copy it over via a super hacky run_command. It won't automatically update though.

Can the header be generated using options or does it require fully custom implementation? If it's just a matter of setting configuration defines I'd just add a bunch of options to toggle each setting.

**Reply by crazybolillo** [reply]
It needs to be physically present, I have gone with the `configuration_file` route and the project is now working. Thanks for the help, hopefully one day I will  be back with a new wrap :)
